packages/@aws-cdk/aws-iam/lib/optimize-statements.ts

```diff
@@ -0,0 +1,286 @@
+/**
+ * Optimizes a list of objects containing IAM policy statements in order to reduce the total size of the rendered JSON
+ * object. The following optimizations are performed:
+ * - Canonicalization of globs in (Not)Resource & (Not)Action: removal entries that are matched by another entry's glob
+ * - De-duplication: removal of statements that have no SID, and are subsets of another statement
+ * - Merging: merging of statements that have no SID, and differ only in either ``Resource``
+ *
+ * Additionally, normalization is performed:
+ * - (Not)Resource & (Not)Action are sorted in alphanumerical order (of the JSON object, if the values are not String)
+ *
+ * @param statements the list of rendered statements to optimize. Those may contain CloudFormation intrinsics.
+ *
+ * @returns a list of equivalents (and hopefylly smaller) statements.
+ */
+export function optimizeStatements(statements: any[]): any[] {
+  const parsed = statements.map(statement => new Statement(statement));
+  return _deduplicate(parsed).map(statement => statement.toJson());
+}
+
+/**
+ * A Policy statement, with convenience functions make it easier to de-duplicate, normalize, ...
+ */
+class Statement implements Deduplicable, Mergeable {
+  private readonly sid?: string;
+
+  private readonly effect: 'Allow' | 'Deny';
+
+  private readonly actions: Glob[];
+  private readonly resources: Glob[];
+
+  private readonly principal: unknown;
+
+  private readonly condition: { [operator: string]: unknown };
+
+  constructor(statement: any) {
+    this.sid = statement.Sid;
+
+    if (statement.Effect !== 'Allow' && statement.Effect !== 'Deny') {
+      throw new Error(`Illegal policy statement effect: ${JSON.stringify(statement.Effect)}`);
+    }
+    this.effect = statement.Effect;
+
+    this.actions = _makeGlobs(statement.Action);
+    this.resources = _makeGlobs(statement.Resource);
+
+    this.principal = statement.Principal || {};
+
+    this.condition = statement.Condition || {};
+
+    function _makeGlobs(expressions: unknown): Glob[] {
+      if (!expressions) { return []; }
+      if (!Array.isArray(expressions)) { return _makeGlobs([expressions]); }
```

Comment:
This line could just be

```
if (!Array.isArray(expressions)) { expressions = [expressions]; }
```

Reply:
Yeah well, the TypeScript compiler isn't smart enough to figure out this forces the `unknown` into a `unknown[]`, so it would require abundant hand-casting.

And if you wondered, the use of `unknown` caught so many instances of "inadvertent bad use of `any`" that I kinda want to not switch to `any`.